### PR TITLE
template: error on missing key

### LINKED ISSUE #3462: Template: Fail job on missing keys
### Nomad version
0.7.0-rc3
### Operating system and Environment details
ubuntu 16.04
### Issue
We are working on moving our cronjobs to Nomad. It includes scheduling scala jars using native exec (no containers). The job itself contains template stanza which renders a job config file with keys from consul before execution. If a key is missing the job will get stuck in pending state and prevents other allocations from running. 
However when adding new jobs they seem to be running while older pending jobs doesn't get allocated.
There is also an issue in the UI which is showing the jobs which are stuck in "running" state
 
### Reproduction steps
Adding 100 new jobs with missing key to have running and pending jobs
Waiting until some failing jobs will be allocated 
Add new working jobs without missing consul key to test allocation  

### Error in UI
rv-job    pending    Missing: kv.block(path_to_key/producerConfigs)

### Job file 
```
{
  "Job": {
    "AllAtOnce": false,
    "Constraints": null,
    "CreateIndex": 34811,
    "Datacenters": [
      "us-east-1b",
      "us-east-1d",
      "us-east-1e"
    ],
    "ID": "dev118-rv-job",
    "JobModifyIndex": 34811,
    "Meta": null,
    "ModifyIndex": 34811,
    "Name": "dev118-rv-job",
    "Namespace": "default",
    "ParameterizedJob": null,
    "ParentID": "",
    "Payload": null,
    "Periodic": null,
    "Priority": 50,
    "Region": "global",
    "Stable": false,
    "Status": "pending",
    "StatusDescription": "",
    "Stop": false,
    "SubmitTime": 1509292961781137149,
    "TaskGroups": [
      {
        "Constraints": null,
        "Count": 1,
        "EphemeralDisk": {
          "Migrate": false,
          "SizeMB": 300,
          "Sticky": false
        },
        "Meta": null,
        "Name": "rv-job",
        "RestartPolicy": {
          "Attempts": 0,
          "Delay": 25,
          "Interval": 600000000000,
          "Mode": "fail"
        },
        "Tasks": [
          {
            "Artifacts": [
              {
                "GetterMode": "file",
                "GetterOptions": null,
                "GetterSource": "http://localhost:8500/v1/kv/nomad-jobs/dev42/rv-job/sonic-templates/c5cbd9b297af0cecc23c02dbba1440fe6e4b182a?raw",
                "RelativeDest": "local/rv-job.conf.tpl"
              },
              {
                "GetterMode": "any",
                "GetterOptions": null,
                "GetterSource": "s3::https://s3.amazonaws.com/file.jar",
                "RelativeDest": "local/"
              }
            ],
            "Config": {
              "command": "/usr/bin/java",
              "args": [
                "-noverify",
                "-Dlogback.configurationFile=local/logback.xml",
                "-DAPPNAME=${appname}",
                "-Xms8192m",
                "-Xmx8192m",
                "-DBRANCH=${appenv}",
                "-Djava.library.path=/var/lib/sonic/lib",
                "-Djava.io.tmpdir=/tmp/",
                "-Dconfig.file=local/${appname}.conf",
                "-cp",
                "local/${commit}.jar",
                "com.supersonic.allocator.AllocatorUpdateProcessMain",
                "rv"
              ]
            },
            "Constraints": null,
            "DispatchPayload": null,
            "Driver": "exec",
            "Env": {
              "appname": "rv-job",
              "submitted": "2017-10-29T16:01:09Z",
              "appenv": "dev42",
              "commit": "c5cbd9b297af0cecc23c02dbba1440fe6e4b182a"
            },
            "KillTimeout": 5000000000,
            "Leader": false,
            "LogConfig": {
              "MaxFileSizeMB": 10,
              "MaxFiles": 10
            },
            "Meta": null,
            "Name": "rv-job",
            "Resources": {
              "CPU": 2048,
              "DiskMB": 0,
              "IOPS": 0,
              "MemoryMB": 8192,
              "Networks": null
            },
            "Services": null,
            "ShutdownDelay": 0,
            "Templates": [
              {
                "ChangeMode": "restart",
                "ChangeSignal": "",
                "DestPath": "local/rv-job.conf",
                "EmbeddedTmpl": "",
                "Envvars": false,
                "LeftDelim": "{{",
                "Perms": "0644",
                "RightDelim": "}}",
                "SourcePath": "local/rv-job.conf.tpl",
                "Splay": 5000000000,
                "VaultGrace": 15000000000
              }
            ],
            "User": "",
            "Vault": null
          }
        ],
        "Update": null
      }
    ],
    "Type": "batch",
    "Update": {
      "AutoRevert": false,
      "Canary": 0,
      "HealthCheck": "",
      "HealthyDeadline": 0,
      "MaxParallel": 0,
      "MinHealthyTime": 0,
      "Stagger": 0
    },
    "VaultToken": "",
    "Version": 0
  }
}
```

### PR DESCRIPTION
**Pull request by DerekStrickland**
This PR adds the consul-template configuration option `error_on_missing_key`. This allows jobspec authors to specify that a template should fail if it references a struct or map key that does not exist. The default value is `false` and should be fully backward compatible.

### CODE REVIEW

File: website/content/docs/job-specification/template.mdx [hunk 1]
```diff
@@ -77,6 +77,9 @@ refer to the [Learn Go Template Syntax][gt_learn] Learn guide.
   validation error. Setting `env` when the `change_mode` is `noop` is
   permitted but will not update the environment variables in the task.
 
+- `error_on_missing_key` `(bool: false)` - Specifies how the template behaves
+  when attempting to index a struct or map key that does not exist.
```
Comment (DerekStrickland) [inline website/content/docs/job-specification/template.mdx:81]:
```suggestion
  when attempting to index a struct or map key that does not exist. If set to true, the task will fail if a template tries to access a key that doesn't exist in a struct or map. If set to false, an empty value will be used in the template instead.
```

@lgfa29  How's this look?

**Comment by rodrigol-chan** [comment]
Thank you for working on this, @DerekStrickland! I've been bitten by this issue. Do you plan on working on this? If not, I can pick this up.

File: jobspec2/parse_test.go [hunk 2]
```diff
@@ -1051,3 +1051,23 @@ func TestWaitConfig(t *testing.T) {
...
+		Body:    hclBytes,
+		AllowFS: false,
+	})
+
+	require.NoError(t, err)
+
+	tmpl := job.TaskGroups[0].Tasks[0].Templates[0]
+	require.NotNil(t, tmpl)
+	require.NotNil(t, tmpl.ErrMissingKey)
+	require.Equal(t, true, *tmpl.ErrMissingKey)
```
[the remark this answers is not in the document]
Comment (angrycub) [reply jobspec2/parse_test.go:1072]: Tested with pointer, doesn't work; so we still need the NotNil check, but agree 100% on switching the test to `require.True`

**Comment by lgfa29** [comment]
> Thank you for working on this, @DerekStrickland! I've been bitten by this issue. Do you plan on working on this? If not, I can pick this up.

Hi @rodrigol-chan 👋 

Could you explain the scenario you had where this would be useful?

We paused work on this because this `consul-template` config didn't actually do what we expect it to do.

**Comment by rodrigol-chan** [comment]
I'm using the Vault integration to template some secrets into a file. Unfortunately, I forgot to add all of the key-value pairs and instead of a nice Nomad error telling me a key-value pair does not exist, I had to figure out that `<no value>` is being produced instead. It wasn't entirely obvious because the secret is salted and hashed, and only then put into a file, but it would be nice to have an explicit failure.

As I understand it, if the whole secret does not exist, it already fails to launch the task. But if a key in the secret does not exist, it's silently templated as the wrong value.

**Comment by angrycub** [comment]
Closing in favor of #15141

**Pull request closed without merging**